### PR TITLE
btcec: optimize square root using fieldVal

### PR DESCRIPTION
**Pull request by cfromknecht**
As of #1193, decompressPoint now
validates that the point is on the curve. The x and y coordinates are
also implicitly <= `P`, since the modular reduction is applied to both
before the method returns. The checks are moved so that they are still
applied when parsing an uncompressed pubkey, as the checks are not
redundant in that path.

We also optimize the decompressPoint subroutine, used in extracting
compressed pubkeys and performing pubkey recovery. We do so by replacing
the use of big.Int.Exp with with square-and-multiply exponentiation of
btcec's more optimized fieldVals, reducing the overall latency and
memory requirements of decompressPoint.

Instead of operating on bits of Q = (P+1)/4, the exponentiation applies
the square-and-multiply operations on full bytes of Q.  Compared to the
original speedup. Compared the bit-wise version, the improvement is
roughly 10%.

A new pair fieldVal methods called Sqrt and SqrtVal are added, which
applies the square-and-multiply exponentiation using precomputed
byte-slice of the value Q.

Comparison against big.Int sqrt and SAM sqrt over bytes of Q:

```
benchmark                            old ns/op     new ns/op     delta
BenchmarkParseCompressedPubKey-8     35545         23119         -34.96%

benchmark                            old allocs     new allocs     delta
BenchmarkParseCompressedPubKey-8     35             6            -82.86%

benchmark                            old bytes     new bytes     delta
BenchmarkParseCompressedPubKey-8     2777          256           -90.78%
```

Finally the `*KoblitzCurve.QPlus1Div4()` method is deprecated, and superseded
by `*KoblitzCurve.Q()`, to convey that the returned value is `Q = (P+1)/4`, and
not `(((P+1)/4)+1)/4) `.

### CODE REVIEW

File: btcec/btcec.go [hunk 1]
```diff
@@ -36,10 +36,21 @@ var (
 // interface from crypto/elliptic.
 type KoblitzCurve struct {
 	*elliptic.CurveParams
-	q         *big.Int
+
+	// q is the value (P+1)/4 use to compute the square root of field
```
Comment (yashbhutwala) [inline btcec/btcec.go]:
```suggestion
	// q is the value (P+1)/4, used to compute the square root of field
```

File: btcec/btcec.go [hunk 2]
```diff
@@ -879,12 +890,33 @@ func (curve *KoblitzCurve) ScalarBaseMult(k []byte) (*big.Int, *big.Int) {
 	return curve.fieldJacobianToBigAffine(qx, qy, qz)
 }
 
-// QPlus1Div4 returns the Q+1/4 constant for the curve for use in calculating
-// square roots via exponention.
+// QPlus1Div4 returns the (P+1)/4 constant for the curve for use in calculating
+// square roots via exponentiation.
+//
+// DEPRECATED: The actual value returned is (P+1)/4, where as the original
+// method name implies that this value is ((P+1)/4)+1)/4. This method is kept to
```
Comment (yashbhutwala) [inline btcec/btcec.go]:
```suggestion
// method name implies that this value is (((P+1)/4)+1)/4. This method is kept to
```

File: btcec/pubkey.go [hunk 3]
```diff
@@ -22,41 +22,139 @@ func isOdd(a *big.Int) bool {
...
+	//
+	//   Q = 3fffffffffffffffffffffffffffffffffffffffffffffffffffffffbfffff0c
+	//
+	// We can unroll the typical bit-wise interpretation of the
+	// exponentiation algorithm above to instead operate on bytes.
+	// This reduces the number of comparisons from 256*3 = 768 in the
+	// prior bit-wise algorithm, down to (32*2)+1+2+2 = 69, reducing the
+	// overhead of failed branch predictions and additional comparisons.
+	//
+	// Since the there are only 4 unique bytes of Q, this keeps the jump
```
Comment (yashbhutwala) [inline btcec/pubkey.go]:
```suggestion
	// Since there are only 4 unique bytes of Q, this keeps the jump
```

**Comment by cfromknecht** [comment]
thanks @yashbhutwala, comments addressed :)

File: btcec/field.go [hunk 4]
```diff
@@ -1221,3 +1235,118 @@ func (f *fieldVal) Inverse() *fieldVal {
...
+	//
+	// The final value r_0 = x^n.
+	//
+	// See https://en.wikipedia.org/wiki/Exponentiation_by_squaring for more
+	// details.
+	//
+	// This can be further optimized, by observing that the value of Q in
+	// secp256k1 has the value:
+	//
+	//   Q = 3fffffffffffffffffffffffffffffffffffffffffffffffffffffffbfffff0c
```
Comment (Roasbeef) [inline btcec/field.go]: This is dope! Can really see how this presents the savings compared to an earlier version that _did_ do the bitwise comparison. Nice having a constant which has such a large Hamming weight.

**Pull request merged**